File: PgManager/Services/TenantService.cs
using Microsoft.EntityFrameworkCore;
using PgManager.Data;
using PgManager.DTOs.Tenant;
using PgManager.Entities;
using PgManager.Services.Interfaces;

namespace PgManager.Services
{
    public class TenantService : ITenantService
    {
        private readonly ApplicationDbContext _context;

        public TenantService(ApplicationDbContext context)
        {
            _context = context;
        }

        // Helper: build TenantDto from entity — rent and sharing type come from Room
        private static TenantDto ToDto(Tenant t) => new TenantDto
        {
            Id = t.Id,
            Name = t.Name,
            PhoneNumber = t.PhoneNumber,
            SharingType = t.Room?.SharingType.ToString() ?? "",
            RoomId = t.RoomId,
            RoomNumber = t.Room?.RoomNumber ?? 0,
            RentAmount = t.Room?.RentPerBed ?? 0,
            AdvanceAmount = t.AdvanceAmount,
            JoinDate = t.JoinDate,
            LastPaidDate = t.LastPaidDate,
            IsActive = t.IsActive,
            DaysSinceLastPayment = t.DaysSinceLastPayment,
            IsOverdue = t.IsOverdue,
            DueAmount = t.DueAmount,
            CurrentDue = t.CurrentDue,
            MonthsElapsed = t.MonthsElapsed,
            CreatedAt = t.CreatedAt
        };

        public async Task<(bool Success, string Message, IEnumerable<TenantDto>? Data, List<string>? Errors)> GetAllTenantsAsync(
            int userId,
            string? searchTerm = null,
            int? sharingType = null,
            bool? isActive = null)
        {
            try
            {
                var query = _context.Tenants
                    .Where(t => t.UserId == userId)
                    .Include(t => t.Room)
                    .AsQueryable();

                if (!string.IsNullOrWhiteSpace(searchTerm))
                {
                    query = query.Where(t => t.Name.Contains(searchTerm) || t.PhoneNumber.Contains(searchTerm));
                }

                // Filter by sharing type — derived from room
                if (sharingType.HasValue)
                {
                    query = query.Where(t => (int)t.Room.SharingType == sharingType.Value);
                }

                if (isActive.HasValue)
                {
                    query = query.Where(t => t.IsActive == isActive.Value);
                }

                var tenants = await query.ToListAsync();

                var sortedTenants = tenants
                    .OrderByDescending(t => t.IsActive)
                    .ThenByDescending(t => t.DueAmount)
                    .Select(ToDto)
                    .ToList();

                return (true, "Tenants retrieved successfully", sortedTenants, null);
            }
            catch (Exception ex)
            {
                return (false, "Error retrieving tenants", null, new List<string> { ex.Message });
            }
        }

        public async Task<(bool Success, string Message, TenantDto? Data, List<string>? Errors)> GetTenantByIdAsync(int userId, int id)
        {
            try
            {
                var tenant = await _context.Tenants
                    .Where(t => t.UserId == userId)
                    .Include(t => t.Room)
                    .FirstOrDefaultAsync(t => t.Id == id);

                if (tenant == null)
                    return (false, "Tenant not found", null, new List<string> { "Tenant does not exist" });

                return (true, "Tenant retrieved successfully", ToDto(tenant), null);
            }
            catch (Exception ex)
            {
                return (false, "Error retrieving tenant", null, new List<string> { ex.Message });
            }
        }

        public async Task<(bool Success, string Message, IEnumerable<TenantDto>? Data, List<string>? Errors)> GetOverdueTenantsAsync(int userId)
        {
            try
            {
                var tenants = await _context.Tenants
                    .Where(t => t.UserId == userId)
                    .Include(t => t.Room)
                    .Where(t => t.IsActive)
                    .ToListAsync();

                var overdueTenants = tenants
                    .Where(t => t.IsOverdue)
                    .OrderByDescending(t => t.DaysSinceLastPayment)
                    .Select(ToDto)
                    .ToList();

                return (true, "Overdue tenants retrieved successfully", overdueTenants, null);
            }
            catch (Exception ex)
            {
                return (false, "Error retrieving overdue tenants", null, new List<string> { ex.Message });
            }
        }

        public async Task<(bool Success, string Message, TenantDto? Data, List<string>? Errors)> CreateTenantAsync(int userId, CreateTenantDto dto)
        {
            try
            {
                // Verify room exists and belongs to user
                var room = await _context.Rooms
                    .Where(r => r.UserId == userId)
                    .Include(r => r.Tenants.Where(t => t.IsActive))
                    .FirstOrDefaultAsync(r => r.Id == dto.RoomId);

                if (room == null)
                    return (false, "Room not found", null, new List<string> { "Room does not exist" });

                // Check capacity
                var activeTenantCount = room.Tenants.Count(t => t.IsActive);
                if (activeTenantCount >= room.TotalBeds)
                    return (false, "Room is full", null, new List<string> { $"Room is at full capacity ({activeTenantCount}/{room.TotalBeds} beds occupied)" });

                var tenant = new Tenant
                {
                    UserId = userId,
                    Name = dto.Name,
                    PhoneNumber = dto.PhoneNumber,
                    RoomId = dto.RoomId,
                    AdvanceAmount = dto.AdvanceAmount,
                    JoinDate = dto.JoinDate,
                    LastPaidDate = dto.LastPaidDate,
                    IsActive = dto.IsActive,
                    DueAmount = dto.DueAmount,
                    CreatedAt = DateTime.UtcNow
                };

                _context.Tenants.Add(tenant);
                await _context.SaveChangesAsync();

                // Reload to get room info
                await _context.Entry(tenant).Reference(t => t.Room).LoadAsync();

                return (true, "Tenant created successfully", ToDto(tenant), null);
            }
            catch (Exception ex)
            {
                return (false, "Error creating tenant", null, new List<string> { ex.Message });
            }
        }

        public async Task<(bool Success, string Message, TenantDto? Data, List<string>? Errors)> UpdateTenantAsync(int userId, int id, UpdateTenantDto dto)
        {
            try
            {
                var tenant = await _context.Tenants
                    .Where(t => t.UserId == userId)
                    .Include(t => t.Room)
                    .FirstOrDefaultAsync(t => t.Id == id);

                if (tenant == null)
                    return (false, "Tenant not found", null, new List<string> { "Tenant does not exist" });

                var oldRoomId = tenant.RoomId;
                var wasActive = tenant.IsActive;

                // Handle room change
                if (dto.RoomId.HasValue && dto.RoomId.Value != tenant.RoomId)
                {
                    var newRoom = await _context.Rooms
                        .Include(r => r.Tenants.Where(t => t.IsActive))
                        .FirstOrDefaultAsync(r => r.Id == dto.RoomId.Value);

                    if (newRoom == null)
                        return (false, "New room not found", null, new List<string> { "New room does not exist" });

                    var newRoomActiveTenants = newRoom.Tenants.Count(t => t.IsActive);
                    if (newRoomActiveTenants >= newRoom.TotalBeds)
                        return (false, "New room is full", null, new List<string> { $"New room is at full capacity ({newRoomActiveTenants}/{newRoom.TotalBeds} beds occupied)" });

                    var oldRoom = await _context.Rooms.FindAsync(oldRoomId);
                    if (oldRoom != null && tenant.IsActive)
                    {
                        oldRoom.OccupiedBeds--;
                        oldRoom.UpdatedAt = DateTime.UtcNow;
                    }

                    if (tenant.IsActive)
                    {
                        newRoom.OccupiedBeds++;
                        newRoom.UpdatedAt = DateTime.UtcNow;
                    }

                    tenant.RoomId = dto.RoomId.Value;
                }

                // Handle IsActive change
                if (dto.IsActive.HasValue && dto.IsActive.Value != wasActive)
                {
                    var room = await _context.Rooms.FindAsync(tenant.RoomId);
                    if (room != null)
                    {
                        if (dto.IsActive.Value && !wasActive)
                        {
                            if (room.OccupiedBeds >= room.TotalBeds)
                                return (false, "Room is full", null, new List<string> { "Cannot activate tenant - no available beds" });
                            room.OccupiedBeds++;
                        }
                        else if (!dto.IsActive.Value && wasActive)
                        {
                            room.OccupiedBeds--;
                        }
                        room.UpdatedAt = DateTime.UtcNow;
                    }
                    tenant.IsActive = dto.IsActive.Value;
                }

                if (!string.IsNullOrWhiteSpace(dto.Name))
                    tenant.Name = dto.Name;

                if (!string.IsNullOrWhiteSpace(dto.PhoneNumber))
                    tenant.PhoneNumber = dto.PhoneNumber;

                if (dto.AdvanceAmount.HasValue)
                    tenant.AdvanceAmount = dto.AdvanceAmount.Value;

                if (dto.JoinDate.HasValue)
                    tenant.JoinDate = dto.JoinDate.Value;

                if (dto.LastPaidDate.HasValue)
                {
                    if (tenant.LastPaidDate != dto.LastPaidDate.Value)
                        tenant.DueAmount = 0;
                    tenant.LastPaidDate = dto.LastPaidDate.Value;
                }

                if (dto.DueAmount.HasValue)
                    tenant.DueAmount = dto.DueAmount.Value;

                tenant.UpdatedAt = DateTime.UtcNow;
                await _context.SaveChangesAsync();

                await _context.Entry(tenant).Reference(t => t.Room).LoadAsync();

                return (true, "Tenant updated successfully", ToDto(tenant), null);
            }
            catch (Exception ex)
            {
                return (false, "Error updating tenant", null, new List<string> { ex.Message });
            }
        }

        public async Task<(bool Success, string Message, TenantDto? Data, List<string>? Errors)> UpdatePaymentAsync(int userId, int id, UpdatePaymentDto updatePaymentDto)
        {
            try
            {
                var tenant = await _context.Tenants
                    .Where(t => t.UserId == userId)
                    .Include(t => t.Room)
                    .FirstOrDefaultAsync(t => t.Id == id);

                if (tenant == null)
                    return (false, "Tenant not found", null, new List<string> { "Tenant does not exist" });

                if (updatePaymentDto.PaidAmount <= 0)
                    return (false, "Invalid payment", null, new List<string> { "Payment amount must be greater than 0" });

                var totalDueBeforePayment = tenant.CurrentDue;

                if (updatePaymentDto.PaidAmount > totalDueBeforePayment)
                    return (false, "Overpayment not allowed", null,
                        new List<string> { $"Payment amount (₹{updatePaymentDto.PaidAmount}) cannot exceed current due (₹{totalDueBeforePayment})" });

                tenant.LastPaidDate = updatePaymentDto.PaymentDate;
                tenant.DueAmount = totalDueBeforePayment - updatePaymentDto.PaidAmount;
                tenant.UpdatedAt = DateTime.UtcNow;

                await _context.SaveChangesAsync();

                return (true, "Payment updated successfully", ToDto(tenant), null);
            }
            catch (Exception ex)
            {
                return (false, "Error updating payment", null, new List<string> { ex.Message });
            }
        }

        public async Task<(bool Success, string Message, List<string>? Errors)> DeleteTenantAsync(int userId, int id)
        {
            try
            {
                var tenant = await _context.Tenants
                    .Where(t => t.UserId == userId)
                    .Include(t => t.Room)
                    .FirstOrDefaultAsync(t => t.Id == id);

                if (tenant == null)
                    return (false, "Tenant not found", new List<string> { "Tenant does not exist" });

                if (tenant.IsActive)
                {
                    var room = await _context.Rooms.FindAsync(tenant.RoomId);
                    if (room != null)
                    {
                        room.OccupiedBeds--;
                        room.UpdatedAt = DateTime.UtcNow;
                    }
                }

                _context.Tenants.Remove(tenant);
                await _context.SaveChangesAsync();

                return (true, "Tenant deleted successfully", null);
            }
            catch (Exception ex)
            {
                return (false, "Error deleting tenant", new List<string> { ex.Message });
            }
        }
    }
}

File: PgManagerBackend/src/services/tenantService.ts
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

export class TenantService {
  async getAllTenants(userId: number, searchTerm?: string, sharingType?: number, isActive?: boolean) {
    try {
      const tenants = await prisma.tenant.findMany({
        where: {
          userId,
          ...(searchTerm && {
            OR: [
              { name: { contains: searchTerm, mode: 'insensitive' } },
              { phoneNumber: { contains: searchTerm } },
            ],
          }),
          ...(isActive !== undefined && { isActive }),
        },
        include: { room: true },
        orderBy: [{ isActive: 'desc' }, { dueAmount: 'desc' }],
      });

      return tenants.map((t) => this.formatTenantDto(t));
    } catch (error) {
      throw error;
    }
  }

  async getTenantById(userId: number, tenantId: number) {
    try {
      const tenant = await prisma.tenant.findFirst({
        where: { id: tenantId, userId },
        include: { room: true },
      });

      if (!tenant) {
        throw new Error('Tenant not found');
      }

      return this.formatTenantDto(tenant);
    } catch (error) {
      throw error;
    }
  }

  async getOverdueTenants(userId: number) {
    try {
      const tenants = await prisma.tenant.findMany({
        where: { userId, isActive: true },
        include: { room: true },
      });

      const overdue = tenants
        .filter((t) => this.isOverdue(t))
        .sort((a, b) => this.getDaysSinceLastPayment(b) - this.getDaysSinceLastPayment(a));

      return overdue.map((t) => this.formatTenantDto(t));
    } catch (error) {
      throw error;
    }
  }

  async createTenant(userId: number, data: {
    name: string;
    phoneNumber: string;
    roomId: number;
    advanceAmount: number;
    joinDate: Date;
    lastPaidDate?: Date;
    isActive?: boolean;
    dueAmount?: number;
  }) {
    try {
      // Validate phone number (10 digits)
      if (!/^\d{10}$/.test(data.phoneNumber)) {
        throw new Error('Phone number must be 10 digits');
      }

      // Check room exists and has capacity
      const room = await prisma.room.findFirst({
        where: { id: data.roomId, userId },
        include: { tenants: { where: { isActive: true } } },
      });

      if (!room) {
        throw new Error('Room not found');
      }

      if (room.tenants.length >= room.totalBeds) {
        throw new Error('Room is full');
      }

      const tenant = await prisma.tenant.create({
        data: {
          userId,
          name: data.name,
          phoneNumber: data.phoneNumber,
          roomId: data.roomId,
          advanceAmount: data.advanceAmount,
          joinDate: data.joinDate,
          lastPaidDate: data.lastPaidDate,
          isActive: data.isActive !== false,
          dueAmount: data.dueAmount || 0,
        },
        include: { room: true },
      });

      return this.formatTenantDto(tenant);
    } catch (error) {
      throw error;
    }
  }

  async updateTenant(userId: number, tenantId: number, data: any) {
    try {
      const tenant = await prisma.tenant.findFirst({
        where: { id: tenantId, userId },
        include: { room: true },
      });

      if (!tenant) {
        throw new Error('Tenant not found');
      }

      const updateData: any = {};

      if (data.name) updateData.name = data.name;
      if (data.phoneNumber) updateData.phoneNumber = data.phoneNumber;
      if (data.advanceAmount !== undefined) updateData.advanceAmount = data.advanceAmount;
      if (data.joinDate) updateData.joinDate = data.joinDate;
      if (data.lastPaidDate !== undefined) {
        updateData.lastPaidDate = data.lastPaidDate;
        if (data.lastPaidDate) updateData.dueAmount = 0;
      }
      if (data.dueAmount !== undefined) updateData.dueAmount = data.dueAmount;

      // Handle room change
      if (data.roomId && data.roomId !== tenant.roomId) {
        const newRoom = await prisma.room.findFirst({
          where: { id: data.roomId, userId },
          include: { tenants: { where: { isActive: true } } },
        });

        if (!newRoom) {
          throw new Error('New room not found');
        }

        if (newRoom.tenants.length >= newRoom.totalBeds && tenant.isActive) {
          throw new Error('New room is full');
        }

        updateData.roomId = data.roomId;
      }

      // Handle isActive change
      if (data.isActive !== undefined && data.isActive !== tenant.isActive) {
        updateData.isActive = data.isActive;
      }

      const updated = await prisma.tenant.update({
        where: { id: tenantId },
        data: updateData,
        include: { room: true },
      });

      return this.formatTenantDto(updated);
    } catch (error) {
      throw error;
    }
  }

  async updatePayment(userId: number, tenantId: number, paidAmount: number, paymentDate: Date) {
    try {
      const tenant = await prisma.tenant.findFirst({
        where: { id: tenantId, userId },
        include: { room: true },
      });

      if (!tenant) {
        throw new Error('Tenant not found');
      }

      if (paidAmount <= 0) {
        throw new Error('Payment amount must be greater than 0');
      }

      const currentDue = this.getCurrentDue(tenant);
      if (paidAmount > currentDue) {
        throw new Error(`Payment cannot exceed current due (₹${currentDue})`);
      }

      const updated = await prisma.tenant.update({
        where: { id: tenantId },
        data: {
          lastPaidDate: paymentDate,
          dueAmount: currentDue - paidAmount,
        },
        include: { room: true },
      });

      return this.formatTenantDto(updated);
    } catch (error) {
      throw error;
    }
  }

  async deleteTenant(userId: number, tenantId: number) {
    try {
      const tenant = await prisma.tenant.findFirst({
        where: { id: tenantId, userId },
      });

      if (!tenant) {
        throw new Error('Tenant not found');
      }

      await prisma.tenant.delete({
        where: { id: tenantId },
      });

      return { message: 'Tenant deleted successfully' };
    } catch (error) {
      throw error;
    }
  }

  // Helper methods
  private formatTenantDto(tenant: any) {
    return {
      id: tenant.id,
      name: tenant.name,
      phoneNumber: tenant.phoneNumber,
      roomId: tenant.roomId,
      roomNumber: tenant.room?.roomNumber || 0,
      sharingType: tenant.room ? this.getSharingType(tenant.room.sharingType) : '',
      rentAmount: tenant.room?.rentPerBed || 0,
      advanceAmount: tenant.advanceAmount,
      joinDate: tenant.joinDate,
      lastPaidDate: tenant.lastPaidDate,
      isActive: tenant.isActive,
      dueAmount: tenant.dueAmount,
      daysSinceLastPayment: this.getDaysSinceLastPayment(tenant),
      isOverdue: this.isOverdue(tenant),
      currentDue: this.getCurrentDue(tenant),
      monthsElapsed: this.getMonthsElapsed(tenant),
      createdAt: tenant.createdAt,
    };
  }

  private getMonthsElapsed(tenant: any): number {
    const referenceDate = tenant.lastPaidDate || tenant.joinDate;
    const today = new Date();
    let months = 0;

    const nextDate = new Date(referenceDate);
    while (nextDate.setMonth(nextDate.getMonth() + 1) <= today) {
      months++;
    }

    return months;
  }

  private getDaysSinceLastPayment(tenant: any): number {
    const referenceDate = tenant.lastPaidDate || tenant.joinDate;
    const today = new Date();
    const timeDiff = today.getTime() - new Date(referenceDate).getTime();
    return Math.floor(timeDiff / (1000 * 60 * 60 * 24));
  }

  private isOverdue(tenant: any): boolean {
    if (!tenant.lastPaidDate) return true;
    return this.getMonthsElapsed(tenant) > 0;
  }

  private getCurrentDue(tenant: any): number {
    const rent = tenant.room?.rentPerBed || 0;
    if (!tenant.lastPaidDate) {
      return tenant.dueAmount + (this.getMonthsElapsed(tenant) + 1) * rent;
    }
    return tenant.dueAmount + this.getMonthsElapsed(tenant) * rent;
  }

  private getSharingType(type: number): string {
    const types: { [key: number]: string } = {
      1: 'Single',
      2: 'Double',
      3: 'Triple',
      4: 'Four',
      5: 'Five',
      6: 'Six',
    };
    return types[type] || 'Unknown';
  }
}
